# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [207]:
import warnings
import pandas as pd
import sqlite3
warnings.simplefilter('ignore')


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [208]:
salesConnection = sqlite3.connect('../sqlite/go_sales.sqlite')
staffConnection = sqlite3.connect('../sqlite/go_staff.sqlite')
crmConnection = sqlite3.connect('../sqlite/go_crm.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [209]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
tables = pd.read_sql_query(sql_query, salesConnection)
tables

name
0            country
1      order_details
2       order_header
3       order_method
4            product
5       product_line
6       product_type
7      retailer_site
8      return_reason
9      returned_item
10      sales_branch
11       sales_staff
12  SALES_TARGETData
13   sqlite_sequence

In [210]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
tables = pd.read_sql_query(sql_query, salesConnection)
tables

name
0            country
1      order_details
2       order_header
3       order_method
4            product
5       product_line
6       product_type
7      retailer_site
8      return_reason
9      returned_item
10      sales_branch
11       sales_staff
12  SALES_TARGETData
13   sqlite_sequence

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [211]:
products = "SELECT * FROM product;"
product = pd.read_sql_query(products, salesConnection)

product['PRODUCTION_COST'] = product['PRODUCTION_COST'].astype(float)
productiekosten_lager100_hoger50 = (product['PRODUCTION_COST'] < 100) & (product['PRODUCTION_COST'] > 50)
product[productiekosten_lager100_hoger50]

PRODUCT_NUMBER INTRODUCTION_DATE PRODUCT_TYPE_CODE  PRODUCTION_COST MARGIN  \
13            110        10-12-2019                20            60.00     .5   
14            111        15-12-2019                20            81.80    .55   
17            114        15-12-2019                21            80.00     .6   
24             17          5-3-2013                 3            60.00    .28   
25             18         15-2-2011                 3            86.00    .33   
28             20          5-3-2013                 3            78.55    .28   
31             23          5-3-2013                 3            65.33    .33   
32             24          5-3-2013                 4            62.50    .17   
56             46         10-1-2021                 7            52.86    .28   
58             48         10-1-2021                 7            55.00     .5   
69             58         10-1-2021                 9            57.13    .28   
71              6          5-3-2013                 1            85.11    .28   
81             69         15-2-2011                11            73.33    .33   
83             70          5-3-2013                11            80.00    .33   
84             71          5-3-2013                11            60.00    .33   
85             72        26-10-2016                11            72.50     .5   
88             75        26-10-2016                12            80.00    .28   
91             78         15-2-2011                13            79.19     .2   
92             79         15-2-2011                13            92.58    .25   
93              8          5-3-2013                 1            75.00    .41   
94             80          5-3-2013                13            94.12     .4   
97             83          5-3-2013                14            56.25    .37   
98             84          5-3-2013                14            78.55    .28   

      PRODUCT_IMAGE LANGUAGE                     PRODUCT_NAME  \
13  P110GE5PT20.jpg       EN                Blue Steel Putter   
14  P111GE5PT20.jpg       EN            Blue Steel Max Putter   
17  P114GE5GA21.jpg       EN              Course Pro Golf Bag   
24    P17CE1SB3.jpg       EN                  Hibernator Lite   
25    P17CE1SB3.jpg       EN                       Hibernator   
28    P20CE1SB3.jpg       EN  Hibernator Self - Inflating Mat   
31    P23CE1SB3.jpg       EN              Hibernator Camp Cot   
32    P24CE1PK4.jpg       EN     Canyon Mule Climber Backpack   
56    P46ME2SY7.jpg       EN          Granite Climbing Helmet   
58    P47ME2SY7.jpg       EN            Husky Harness Extreme   
69    P58ME2TL9.jpg       EN                   Granite Hammer   
71    P06CE1CG1.jpg       EN        TrailChef Deluxe Cook Set   
81   P69PA3EW11.jpg       EN                        Polar Ice   
83   P70PA3EW11.jpg       EN                     Polar Sports   
84   P71PA3EW11.jpg       EN                       Polar Wave   
85   P72PA3EW11.jpg       EN                    Polar Extreme   
88   P75PA3KV12.jpg       EN                     Edge Extreme   
91   P78PA3BN13.jpg       EN                        Seeker 35   
92   P78PA3BN13.jpg       EN                        Seeker 50   
93    P08CE1CG1.jpg       EN           TrailChef Double Flame   
94   P80PA3BN13.jpg       EN                   Seeker Extreme   
97   P83PA3NV14.jpg       EN                   Glacier Deluxe   
98   P84PA3NV14.jpg       EN                      Glacier GPS   

                                          DESCRIPTION TRIAL888  
13  This premium putter has copper mixed with ligh...        T  
14  Putter head is composed from a single piece of...        T  
17  High quality golf bag that includes carrying s...        T  
24  The Hibernator Lite is the perfect summer slee...        T  
25  The Hibernator is a three-season sleeping bag....        T  
28  Mattress is made of extra-thick foam. Mesh stu...        T  
31  Aluminum frame camp cot that is lightw

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [212]:
product['MARGIN'] = product['MARGIN'].astype(float)

marginPercentage = (product['MARGIN'] < .2) | (product['MARGIN'] > .6)
product[marginPercentage]

PRODUCT_NUMBER INTRODUCTION_DATE PRODUCT_TYPE_CODE  PRODUCTION_COST  \
29              21         15-2-2011                 3            29.17   
32              24          5-3-2013                 4            62.50   
100             86         15-2-2011                15             2.00   
101             87         15-2-2011                15             2.00   
102             88         15-2-2011                15             2.00   
103             89         15-2-2011                15             2.50   
105             90         15-2-2011                15             2.50   

     MARGIN   PRODUCT_IMAGE LANGUAGE                  PRODUCT_NAME  \
29     0.17   P21CE1SB3.jpg       EN                Hibernator Pad   
32     0.17   P24CE1PK4.jpg       EN  Canyon Mule Climber Backpack   
100    0.67  P86OP4IR15.jpg       EN             BugShield Natural   
101    0.67  P86OP4IR15.jpg       EN               BugShield Spray   
102    0.70  P86OP4IR15.jpg       EN         BugShield Lotion Lite   
103    0.63  P86OP4IR15.jpg       EN              BugShield Lotion   
105    0.63  P86OP4IR15.jpg       EN             BugShield Extreme   

                                           DESCRIPTION TRIAL888  
29   Warms up quickly for snow camping comfort. The...        T  
32   This pack is perfect for day trips and short h...        T  
100  Contains no chemicals. Repels mosquitoes, blac...        T  
101  Repels black flies, ticks, mosquitoes, chigger...        T  
102  Repels black flies, ticks, mosquitoes, chigger...        T  
103  Repels black flies, ticks, mosquitoes, chigger...        T  
105  Highest amount of the active ingredient, 38%, ...        T

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [213]:
countrys = "SELECT * FROM country;"
country = pd.read_sql_query(countrys, salesConnection)

countryCurrency = country['CURRENCY_NAME'] == 'francs'
country.sort_values(by='COUNTRY')[countryCurrency]

COUNTRY_CODE      COUNTRY LANGUAGE CURRENCY_NAME TRIAL879
7            17      Belgium       EN        francs        T
0             1       France       EN        francs        T
19            8  Switzerland       EN        francs        T

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [214]:
product['MARGIN'] = product['MARGIN'].astype(float)

productIntroDateSort = (product['MARGIN'] > .5) & (product['INTRODUCTION_DATE'])
introDates = product.loc[productIntroDateSort, 'INTRODUCTION_DATE'].unique()

pd.DataFrame(introDates, columns=['INTRODUCTION_DATE'])


INTRODUCTION_DATE
0        10-12-2019
1        18-12-2019
2        27-12-2019
3         13-1-2020
4        15-12-2019
5        26-10-2016
6         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [215]:
retailers = "SELECT * FROM retailer_site;"
retailer = pd.read_sql_query(retailers, salesConnection)

retailerAddress2REGIONNotNull = retailer['ADDRESS2'].notna() & retailer['REGION'].notna()
retailer[retailerAddress2REGIONNotNull][['ADDRESS1', 'CITY']]


ADDRESS1           CITY
2            4830 Princeton    Albuquerque
5        1700 George Street  San Francisco
13        1488 Orleans Road     Charleston
28      2103 Ashbury Street    Los Angeles
29           333 Ustaff Way      San Diego
..                      ...            ...
367      10423 Kenwood Road     Cincinnati
370   332 South Main Street        Hanover
371     795 South Jefferson    Springfield
376      702 North Meridian  Oklahoma City
377  1210 Somerville Avenue    Chattanooga

[78 rows x 2 columns]

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [216]:
countrys = "SELECT * FROM country;"
country = pd.read_sql_query(countrys, salesConnection)

countryCurrency = (country['CURRENCY_NAME'] == 'dollars') | (country['CURRENCY_NAME'] == 'new dollar')
country.sort_values(by='COUNTRY')[countryCurrency]

COUNTRY_CODE        COUNTRY LANGUAGE CURRENCY_NAME TRIAL879
6            16      Australia       EN       dollars        T
15            4         Canada       EN       dollars        T
3            12         Taiwan       EN    new dollar        T
14            3  United States       EN       dollars        T

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [217]:
getCustomerLocation = (retailer['POSTAL_ZONE'].str[0] == 'D') & (retailer['ADDRESS2'].notna())
retailer.loc[getCustomerLocation, ['ADDRESS1', 'ADDRESS2', 'CITY']]


ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [218]:
return_items = "SELECT * FROM returned_item;"
return_item = pd.read_sql_query(return_items, salesConnection)

total_returned_items = return_item['RETURN_QUANTITY']
total_returned_items.count()

706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [219]:
sales_branchs = "SELECT * FROM sales_branch;"
sales_branch = pd.read_sql_query(sales_branchs, salesConnection)

number_of_regions = sales_branch['REGION'].nunique()
number_of_regions

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [220]:
max_product_price = product['PRODUCTION_COST'].max()
min_product_price = product['PRODUCTION_COST'].min()
avg_product_price = product['PRODUCTION_COST'].mean()

max_product_price, min_product_price, avg_product_price

(555.9, 1.0, 89.75460869565217)

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [221]:
getCustomerLocation = retailer['ADDRESS2'].isna()
getCustomerLocation.count()


391

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [222]:
order_details = "SELECT * FROM order_details;"
order_detail = pd.read_sql_query(order_details, salesConnection)

order_detail_discount = order_detail['UNIT_SALE_PRICE'] < order_detail['UNIT_PRICE']
discounted_items = order_detail[order_detail_discount]
discounted_items['UNIT_PRICE'] = pd.to_numeric(discounted_items['UNIT_PRICE'], errors='coerce')
average_cost_price_discounted = discounted_items['UNIT_PRICE'].mean()

average_cost_price_discounted

192.88697889934417

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [223]:
sales_staffi = "SELECT * FROM sales_staff;"
sales_staff = pd.read_sql_query(sales_staffi, staffConnection)

sales_staff.groupby('POSITION_EN')['SALES_STAFF_CODE'].count()

POSITION_EN
Branch Manager                  19
District Manager                 4
General Manager                  1
Level 1 Sales Representative    13
Level 2 Sales Representative    39
Level 3 Sales Representative    24
Regional Manager                 2
Sales Representative             2
Name: SALES_STAFF_CODE, dtype: int64

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [224]:
staff_counts = sales_staff.groupby('WORK_PHONE')['SALES_STAFF_CODE'].count()
filtered_counts = staff_counts[staff_counts > 4]
filtered_counts

WORK_PHONE
+(44) 121 3505267      5
+(61) 03 2982  4242    5
+31 (0)20 692 93 94    5
+33 1 68 94 52 20      5
1 (206) 292-0012       5
1 (305) 557-4810       6
1 (310) 281-5722       5
1 (403) 232-5986       5
1 (416) 493-5595       5
1 (617) 268-6754       6
Name: SALES_STAFF_CODE, dtype: int64

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [225]:
merged_df = pd.merge(retailer, country, left_on='COUNTRY_CODE', right_on='COUNTRY_CODE', how='inner')
sales_branch_location = merged_df[['ADDRESS1', 'CITY']]
nl_sales_branch_location = merged_df['COUNTRY'] == 'Netherlands'
filtered_sales_branch = sales_branch_location[nl_sales_branch_location]
filtered_sales_branch

ADDRESS1          CITY
164                   Beets 36         Beets
165               Perestraat 1      Deventer
166  Laan van Meerdervoort 966      Den Haag
167             Kalfjeslaan 47    Amstelveen
168               Startbaan 11    Amstelveen
169      Noorderspoorsingel 28     Groningen
170              Hogehilweg 13     Amsterdam
171             Spoorstraat 43    Varsseveld
173             Westhavenweg 8     Amsterdam
174              Westzeedijk 4        Muiden
175            Gooiergracht 91         Laren
176            Hoofdstraat 150   Pieterburen
177                Olde Diek 2         Laren
178     Amsterdamsestraatweg 8      Maarssen
179        Krayenhoffstraat 11     Amsterdam
180         Overschiestraat 19     Amsterdam
181              Havenbuurt 12        Marken
182               Duivendaal 1    Wageningen
184              Voorstraat 37  Nieuweschans
185         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [226]:
product_types = "SELECT * FROM product_type;"
product_type = pd.read_sql_query(product_types, salesConnection)

merged_df = pd.merge(product, product_type, left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE', how='inner')
eyewear_type = merged_df['PRODUCT_TYPE_EN'] == 'Eyewear'
result = merged_df.loc[eyewear_type, 'PRODUCT_NAME']
result


80        Polar Sun
81        Polar Ice
83     Polar Sports
84       Polar Wave
85    Polar Extreme
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [229]:
merged_df = pd.merge(sales_staff, sales_branch, left_on='SALES_BRANCH_CODE', right_on='SALES_BRANCH_CODE', how='inner')
unique_adresses_names = merged_df[merged_df['POSITION_EN'] == 'Branch Manager']
unique_adresses_names.loc[:, ['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].drop_duplicates('ADDRESS1')


ADDRESS1 FIRST_NAME      LAST_NAME
7                     Victoria House     Walter         Taylor
11                    3 Albany Court      Polly         Linton
19                 Interleuvenlaan 2      Gracy        Gellens
27                    Schwabentor 35      Frank          Fuchs
30                  Leopoldstraße 36      Fritz         Hirsch
33                   Piazza Duomo, 1      Maria      Iacobucci
38               Singelgravenplein 4       Kick        Kalkman
43               Isafjordsgatan 30 C      Karin      Bergström
44     75, rue du Faubourg St-Honoré      Denis           Pagé
47      7800, 756 - 6th Avenue. S.W.      Sally          White
50                  789 Yonge Street    Georges  Saint-Germain
56            1288 Dorchester Avenue       Alex      Rodriguez
60                   299 Yale Avenue   Samantha          Floyd
65                    10032 NW 186th      Julie          Olsen
69        1288 South Barrington Ave.      Corey         Wright
80               6c, rue de l'Église      Julie        Béchard
84  Prol. Paseo de la Reforma No. 51      Lucía          Reyna
91                  2315 Queen's Ave      Alice         Walter
98             Avenida Paulista, 333     Bayard          Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [144]:
orderheaders = "SELECT * FROM order_header;"
order_header = pd.read_sql_query(orderheaders, salesConnection)

merged_orderheader_salesstaff = pd.merge(order_header, sales_staff, on='SALES_STAFF_CODE')
staff_contains_manager = merged_orderheader_salesstaff['POSITION_EN'].str.contains('Manager')

merged_orderheader_salesstaff[staff_contains_manager][['POSITION_EN', 'ORDER_DATE']]

POSITION_EN  ORDER_DATE
3574  Branch Manager  2022-09-09
3575  Branch Manager  2022-10-31
3576  Branch Manager  2022-12-13
3577  Branch Manager  2022-03-13
3578  Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [22]:
merged_df = pd.merge(product, product_type, left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE', how='inner')
merged_df = pd.merge(merged_df, order_detail, left_on='PRODUCT_NUMBER', right_on='PRODUCT_NUMBER', how='inner')

filtered_df = merged_df[merged_df['QUANTITY'] > 750]
result = filtered_df[['PRODUCT_TYPE_EN', 'PRODUCT_NAME']].drop_duplicates()
result

PRODUCT_TYPE_EN           PRODUCT_NAME
8616               Tents               Star Peg
15975           Lanterns              Firefly 2
37903  Insect Repellents        BugShield Spray
38313  Insect Repellents  BugShield Lotion Lite
38770  Insect Repellents       BugShield Lotion
39618  Insect Repellents      BugShield Extreme
40754          Sunscreen         Sun Shelter 15
41106          Sunscreen         Sun Shelter 30
41681          Sunscreen             Sun Shield

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [30]:
merged_df = pd.merge(product, order_detail, left_on='PRODUCT_NUMBER', right_on='PRODUCT_NUMBER', how='inner')

merged_df['CALCULATION'] = (merged_df['UNIT_PRICE'].astype(float) - merged_df['UNIT_SALE_PRICE'].astype(float)) / merged_df['UNIT_PRICE'].astype(float)
filtered_df = merged_df[merged_df['CALCULATION'] > 0]
result = filtered_df['PRODUCT_NAME']
result

6        TrailChef Water Bag
7        TrailChef Water Bag
8        TrailChef Water Bag
9        TrailChef Water Bag
10       TrailChef Water Bag
                ...         
43031            Aloe Relief
43032            Aloe Relief
43034            Aloe Relief
43036            Aloe Relief
43059            Aloe Relief
Name: PRODUCT_NAME, Length: 14545, dtype: object

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [231]:
return_reasons = "SELECT * FROM return_reason;"
return_reason = pd.read_sql_query(return_reasons, salesConnection)

merge_returnreason_returneditem = pd.merge(return_reason, return_item, on='RETURN_REASON_CODE')
merge_orderdetails_returnreason_returneditem = pd.merge(merge_returnreason_returneditem, order_details, on='ORDER_DETAIL_CODE')

merge_orderdetails_returnreason_returneditem['RETURN_QUANTITY'] = merge_orderdetails_returnreason_returneditem['RETURN_QUANTITY'].astype(float)
merge_orderdetails_returnreason_returneditem['QUANTITY'] = merge_orderdetails_returnreason_returneditem['QUANTITY'].astype(float)

teruggebracht_hoger90procent_query = merge_orderdetails_returnreason_returneditem['RETURN_QUANTITY'] / merge_orderdetails_returnreason_returneditem['QUANTITY'] > .9
teruggebracht_hoger90procent = merge_orderdetails_returnreason_returneditem[teruggebracht_hoger90procent_query]

teruggebracht_hoger90procent.loc[:, ['RETURN_DESCRIPTION_EN']].drop_duplicates('RETURN_DESCRIPTION_EN')

TypeError: Can only merge Series or DataFrame objects, a <class 'str'> was passed

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [66]:
merged_df = pd.merge(product, product_type, left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE', how='inner')

merged_df.groupby('PRODUCT_TYPE_EN')['PRODUCT_NUMBER'].count()

PRODUCT_TYPE_EN
Binoculars               4
Climbing Accessories     7
Cooking Gear            10
Eyewear                  5
First Aid                5
Golf Accessories         4
Insect Repellents        5
Irons                    4
Knives                   5
Lanterns                12
Navigation               4
Packs                    6
Putters                  3
Rope                     4
Safety                   4
Sleeping Bags            7
Sunscreen                5
Tents                    6
Tools                    6
Watches                  5
Woods                    4
Name: PRODUCT_NUMBER, dtype: int64

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [69]:
merged_df = pd.merge(country, retailer, left_on='COUNTRY_CODE', right_on='COUNTRY_CODE', how='inner')

merged_df.groupby('COUNTRY')['RETAILER_SITE_CODE'].count()

COUNTRY
Australia          10
Austria            10
Belgium            10
Brazil              1
Canada             31
China              10
Denmark             6
Finland             5
France             29
Germany            32
Italy              12
Japan              25
Korea              10
Mexico             10
Netherlands        20
Spain               9
Sweden             12
Switzerland        13
Taiwan             10
United Kingdom     25
United States     101
Name: RETAILER_SITE_CODE, dtype: int64

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [94]:
# Filter the DataFrame to include only 'Cooking Gear' products
merged_df = pd.merge(product, product_type, left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE', how='inner')
merged_df = merged_df[merged_df['PRODUCT_TYPE_EN'] == 'Cooking Gear']
cooking_gear_df = pd.merge(merged_df, order_detail, left_on='PRODUCT_NUMBER', right_on='PRODUCT_NUMBER', how='inner')
cooking_gear_df
# Group by product name and calculate the total quantity sold and the average selling price
overview = cooking_gear_df.groupby('PRODUCT_NAME').agg(
    TOTAL_QUANTITY_SOLD=('QUANTITY', 'sum'),
    AVERAGE_SELLING_PRICE=('PRODUCTION_COST', 'mean')
)
overview = overview.sort_values('TOTAL_QUANTITY_SOLD', ascending=False)
overview

TOTAL_QUANTITY_SOLD  AVERAGE_SELLING_PRICE
PRODUCT_NAME                                                         
TrailChef Water Bag                      41770                   4.00
TrailChef Cup                            31834                   5.00
TrailChef Canteen                        31712                   9.22
TrailChef Kitchen Kit                    25964                  15.93
TrailChef Utensils                       15928                  10.00
TrailChef Double Flame                   14510                  75.00
TrailChef Cook Set                       12700                  34.97
TrailChef Single Flame                   10614                  46.38
TrailChef Kettle                          9900                   9.00
TrailChef Deluxe Cook Set                 3744                  85.11

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [232]:
merge_salesstaff_salestarget = pd.merge(sales_staff, sales_target, on='SALES_STAFF_CODE')
mergebat = pd.merge(merge_salesstaff_salestarget, retailer, on='RETAILER_CODE')
erm = mergebat['CITY'] == 'Amsterdam'
mergebat[erm][['FIRST_NAME', 'LAST_NAME', 'SALES_TARGET']].groupby(['FIRST_NAME', 'LAST_NAME'])['SALES_TARGET'].agg({min, max})

max   min
FIRST_NAME LAST_NAME                   
Belinda    Jansen-Velasquez   960  1005
Ellen      Shapiro            997   100
Jan        Haverkamp          997  1001
Maria      Laponder          9993  1009

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [142]:
merged_df = pd.merge(retailer, country, left_on='COUNTRY_CODE', right_on='COUNTRY_CODE', how='inner')

merged_df.groupby('COUNTRY')[['CITY', 'REGION']].count().rename(columns={'CITY': 'Verkoper', 'REGION': 'Klanten'})


Verkoper  Klanten
COUNTRY                          
Australia             10       10
Austria               10        0
Belgium               10        0
Brazil                 1        1
Canada                31       31
China                 10        6
Denmark                6        0
Finland                5        0
France                29        0
Germany               32       32
Italy                 12        0
Japan                 25       25
Korea                 10        0
Mexico                10       10
Netherlands           20        0
Spain                  9        9
Sweden                12        0
Switzerland           13        0
Taiwan                10       10
United Kingdom        25       16
United States        101      101

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [155]:
merged_df = pd.merge(sales_staff, sales_target, left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE', how='left')

# Filter for sales staff who have never sold anything
no_sales = merged_df[merged_df['SALES'].isna() | (merged_df['SALES'] == 0)]

# Select only the first name and last name columns
no_sales = no_sales[['FIRST_NAME', 'LAST_NAME']]

no_sales

KeyError: 'SALES'

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [157]:
average_margin_all = product['MARGIN'].mean()
below_average_margin = product[product['MARGIN'] < average_margin_all]
product_count = below_average_margin.shape[0]
average_margin_below = below_average_margin['MARGIN'].mean()

overview = pd.DataFrame({
    'Count of Products with Below Average Margin': [product_count],
    'Average Margin of These Products': [average_margin_below]
})
overview

Count of Products with Below Average Margin  \
0                                           59   

   Average Margin of These Products  
0                          0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [167]:
merged_df = pd.merge(product, order_detail, left_on='PRODUCT_NUMBER', right_on='PRODUCT_NUMBER', how='inner')


not_returned = merged_df[~merged_df['ORDER_DETAIL_CODE'].isin(return_item['ORDER_DETAIL_CODE'])]

# Filter the products that have a selling price greater than 500
overview = not_returned[not_returned['UNIT_SALE_PRICE'].astype(float) > 500]['PRODUCT_NAME']

overview

1226     Hailstorm Steel Irons
1227     Hailstorm Steel Irons
1228     Hailstorm Steel Irons
1229     Hailstorm Steel Irons
1230     Hailstorm Steel Irons
                 ...          
21159           Husky Rope 200
21160           Husky Rope 200
21161           Husky Rope 200
21162           Husky Rope 200
21163           Husky Rope 200
Name: PRODUCT_NAME, Length: 4288, dtype: object

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [173]:
is_manager = []

for position in sales_staff['POSITION_EN']:
    if 'Manager' in position:
        is_manager.append('Yes')
    else:
        is_manager.append('No')

sales_staff['IS_MANAGER'] = is_manager
overview = sales_staff[['LAST_NAME', 'IS_MANAGER']]
overview

LAST_NAME IS_MANAGER
0    Savolainen         No
1           Kim         No
2          Choi         No
3           Kao         No
4          Fang         No
..          ...        ...
99       Barros         No
100   Guimarães         No
101    Lindholm         No
102    Petersen         No
103    Petersen         No

[104 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [192]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [175]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [200]:
from datetime import date

current_year = date.today().year

kort_in_dienst = []
lang_in_dienst = []

for start_date in sales_staff['DATE_HIRED']:

    years_in_service = current_year - current_year
    
    # Check if the employee has been in service for less than 25 years
    if years_in_service < 25:
        kort_in_dienst.append('Yes')
    else:
        kort_in_dienst.append('No')
    
    # Check if the employee has been in service for 12 years or more
    if years_in_service >= 12:
        lang_in_dienst.append('Yes')
    else:
        lang_in_dienst.append('No')

# Add the lists as new columns in sales_staff
sales_staff['KORT_IN_DIENST'] = kort_in_dienst
sales_staff['LANG_IN_DIENST'] = lang_in_dienst

# Display the DataFrame with only 'LAST_NAME', 'KORT_IN_DIENST', and 'LANG_IN_DIENST' columns
overview = sales_staff[['LAST_NAME', 'KORT_IN_DIENST', 'LANG_IN_DIENST']]

overview

LAST_NAME KORT_IN_DIENST LANG_IN_DIENST
0    Savolainen            Yes             No
1           Kim            Yes             No
2          Choi            Yes             No
3           Kao            Yes             No
4          Fang            Yes             No
..          ...            ...            ...
99       Barros            Yes             No
100   Guimarães            Yes             No
101    Lindholm            Yes             No
102    Petersen            Yes             No
103    Petersen            Yes             No

[104 rows x 3 columns]